docs/Developers/NextGen_Core.rst

```diff
@@ -0,0 +1,196 @@
+NextGen-Core
+============
+
+This document provides a brief overview of coala's NextGen-Core.
+coala's NextGen-Core comes with the promise of eliminating the dependencies
+amongst bears (by removing the queuing mechanism) and better performance
+(by utilizing asynchronous I/O and process pool).
+
+What has changed?
+-----------------
+
+``Global Bears`` are now called ``Project Bears`` and ``Local Bears`` are
+now known as ``File Bears``. Both the ``ProjectBear`` and the ``FileBear``
+classes inherit from the base class ``coalib.core.Bear.py``.
+
+- **ProjectBears:** Performs analysis on the whole codebase and does not
+  parallelize tasks. All the bears formerly known as GlobalBears will now
+  inherit from this class.
+- **FileBears:** Performs analysis on the files provided by the user
+  and parallelizes tasks for each give file. All the bears formerly known as
+  LocalBears will now inherit from this class.
+
+What is new?
+------------
+
+The following new features have been added as a part of the NextGen-Core:
+
+- New bear type: ``DependencyBear``
+- Usage of instance-bound ``Bear.BEAR_DEPS``
+- Usage of instance-bound ``Bear.REQUIREMENTS``
```

Comment:
- Superior caching
- Improved dependency system
- Custom parallelization for bears

The instance-bound `Bear.REQUIREMENTS` isn't merged yet, and still has a conceptual problem. Let's leave it for now :+1:

Reply:
- Easier interface: Just do `core.run(bears)`
- Official support for virtual files (everything that happens before `core.run` isn't relevant, the bears have to point to the right file data objects when run, whether they are real files or virtual ones)